common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -606,13 +644,18 @@ private ConscryptSession provideSession() {
                 return SSLNullSession.getNullSession();
             }
             return activeSession;
+        } finally {
+            sslLock.unlock();
         }
     }
 
     private ConscryptSession provideHandshakeSession() {
-        synchronized (ssl) {
+        sslLock.lock();
+        try {
             return state == STATE_HANDSHAKE_STARTED ? activeSession
                 : SSLNullSession.getNullSession();
+        } finally {
+            sslLock.unlock();
         }
     }
 
```

Comment:
`provideAfterHandshakeSession()` also needs locking (my bad!)

Reply:
Happy to fix it up. We may be able to add a `compileOnly("com.google.errorprone:error_prone_annotations:$errorproneVersion")` to take advantage of `@GuardedBy` checks without introducing a dependency for consumers.

Reply:
The `exportKeyingMaterial` is interesting because it locks over the `state` check, but not the `NativeSsl` interaction: https://github.com/google/conscrypt/blob/b14310defd12567f4369334751bc01c1c43e79e1/common/src/main/java/org/conscrypt/ConscryptEngine.java#L1839-L1849